sdk/rust/nym-sdk/src/stream_wrapper/mixnet_stream_wrapper_ipr.rs

```diff
@@ -0,0 +1,777 @@
+        let stream = MixStream::new(None, Recipient::from(ipr_address)).await;
+
+        Ok(Self {
+            stream,
+            ipr_address,
+            listener: IprListener::new(),
+            allocated_ips: None,
+            connection_state: ConnectionState::Disconnected,
+        })
+    }
+
+    pub fn nym_address(&self) -> &Recipient {
+        self.stream.client.nym_address()
+    }
+
+    async fn send_ipr_request(&mut self, request: IpPacketRequest) -> Result<(), Error> {
+        let request_bytes = request.to_bytes()?;
+        self.stream.write_bytes(&request_bytes).await
+    }
+
+    pub async fn connect_tunnel(&mut self) -> Result<IpPair, Error> {
+        if self.connection_state != ConnectionState::Disconnected {
+            return Err(Error::IprStreamClientAlreadyConnectedOrConnecting);
+        }
+
+        self.connection_state = ConnectionState::Connecting;
+        info!("Connecting to IP packet router: {}", self.ipr_address);
+
+        match self.connect_inner().await {
+            Ok(ip_pair) => {
+                self.allocated_ips = Some(ip_pair);
+                self.connection_state = ConnectionState::Connected;
+                info!(
+                    "Connected to IPv4: {}, IPv6: {}",
+                    ip_pair.ipv4, ip_pair.ipv6
+                );
+                Ok(ip_pair)
+            }
+            Err(e) => {
+                self.connection_state = ConnectionState::Disconnected;
+                error!("Failed to connect: {:?}", e);
+                Err(e)
+            }
+        }
+    }
+
+    async fn connect_inner(&mut self) -> Result<IpPair, Error> {
+        let (request, request_id) = IpPacketRequest::new_connect_request(None);
+        debug!("Sending connect request with ID: {}", request_id);
+
+        self.send_ipr_request(request).await?;
+        self.listen_for_connect_response(request_id).await
+    }
+
+    async fn listen_for_connect_response(&mut self, request_id: u64) -> Result<IpPair, Error> {
+        let timeout = tokio::time::sleep(IPR_CONNECT_TIMEOUT);
+        tokio::pin!(timeout);
+
+        let mut framed = FramedRead::new(&mut self.stream, ReconstructedMessageCodec {});
+
+        loop {
+            tokio::select! {
+                _ = &mut timeout => {
+                    return Err(Error::IPRConnectResponseTimeout);
+                }
+                frame = framed.next() => {
+                    match frame {
+                        None => {
+                            return Err(Error::IPRClientStreamClosed);
+                        }
+                        Some(Err(e)) => {
+                            return Err(Error::MessageRecovery(e));
+                        }
+                        Some(Ok(reconstructed)) => {
+                            if let Err(e) = check_ipr_message_version(&reconstructed) {
+                                error!("Version check failed: {}", e);
+                                continue;
```

Comment:
if version check failed, is there any point in continuing given we won't be able to talk with the IPR?

Reply:
good point